File: frontend/src/utils/pipelineStage.ts
/** Display pipeline step — terminal statuses plus in-progress sub-stages for `ingested`. */
export type PipelineStage =
  | 'queued'
  | 'ingesting'
  | 'awaiting_analyze'
  | 'awaiting_fingerprint'
  | 'awaiting_tag'
  | 'awaiting_route'
  | 'ready'
  | 'review'
  | 'duplicate'
  | 'failed'
  | 'archived'

/** Dashboard chips in worker order (archived omitted from main row). */
export const PIPELINE_STAGE_ORDER: PipelineStage[] = [
  'queued',
  'ingesting',
  'awaiting_analyze',
  'awaiting_fingerprint',
  'awaiting_tag',
  'awaiting_route',
  'ready',
  'review',
  'duplicate',
  'failed',
]

const STAGE_LABELS: Record<PipelineStage, string> = {
  queued: 'Queued',
  ingesting: 'Ingesting',
  awaiting_analyze: 'Awaiting analyze',
  awaiting_fingerprint: 'Analyzed',
  awaiting_tag: 'Awaiting tag',
  awaiting_route: 'Awaiting route',
  ready: 'Ready',
  review: 'Needs review',
  duplicate: 'Duplicate',
  failed: 'Failed',
  archived: 'Archived',
}

const STAGE_COLORS: Record<PipelineStage, string> = {
  queued: 'warning',
  ingesting: 'info',
  awaiting_analyze: 'primary',
  awaiting_fingerprint: 'primary',
  awaiting_tag: 'primary',
  awaiting_route: 'primary',
  ready: 'success',
  review: 'warning',
  duplicate: 'secondary',
  failed: 'error',
  archived: 'default',
}

export function pipelineStageLabel(stage: PipelineStage): string {
  return STAGE_LABELS[stage] ?? stage
}

export function pipelineStageColor(stage: PipelineStage): string {
  return STAGE_COLORS[stage] ?? 'default'
}

export function isPipelineStage(value: string): value is PipelineStage {
  return value in STAGE_LABELS
}

File: backend/app/utils/pipeline_stage.py
"""Derive a display pipeline step from track status and progress fields."""

from app.models.enums import TrackStatus
from app.models.track import Track


def resolve_pipeline_stage(track: Track, *, has_fingerprint: bool) -> str:
    """
    Map a track to the next pipeline step it is waiting on.

    Terminal statuses mirror ``TrackStatus``. ``INGESTED`` tracks show where they
    are in analyze → fingerprint → tag → route based on completed fields, not
    retained DB rows alone (e.g. reanalyze keeps fingerprint rows but still shows
    ``awaiting_fingerprint`` until analyze has run again).
    """
    status = track.status
    if status == TrackStatus.PROCESSING:
        return "ingesting"
    if status == TrackStatus.QUEUED:
        return "queued"
    if status == TrackStatus.READY:
        return "ready"
    if status == TrackStatus.REVIEW:
        return "review"
    if status == TrackStatus.DUPLICATE:
        return "duplicate"
    if status == TrackStatus.FAILED:
        return "failed"
    if status == TrackStatus.ARCHIVED:
        return "archived"

    if track.integrated_lufs is None:
        return "awaiting_analyze"
    if not has_fingerprint:
        return "awaiting_fingerprint"
    if track.tagged_at is None:
        return "awaiting_tag"
    return "awaiting_route"
